File: leetcode/c++/6_zigzag_conversion.cpp
// https://leetcode.cn/problems/zigzag-conversion/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows < 2)
            return s;
        vector<string> res(numRows);
        int i = 0, flag = -1;
        for (char ch : s) {
           res[i].push_back(ch);
           if (i == 0 || i == numRows - 1)
            flag = - flag;
            i += flag;
        }
        string res_string;
        for (i = 0; i < numRows; i++) 
            res_string += res[i];
        return res_string;
    }
};

// 时间复杂度为和空间复杂度为O(n)

File: leetcode/golang/6_zigzag_conversion.go
func convert(s string, numRows int) string {
    if numRows < 2 {
        return s
    }
    res := make([]string, numRows)
    i, flag := 0, -1
    for j := 0; j < len(s); j++ {
        res[i] += string(s[j])
        if i == 0 || i == numRows -1 {
            flag = -flag
        }
        i += flag
    }
    var resString string
    for i = 0; i < numRows; i++ {
        resString += res[i]
    } 
    return resString
}